src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -277,6 +283,29 @@ private SocketAsyncEngine()
                     throw new InternalException(err);
                 }
 
+                Interop.Sys.AioContext aioContext = default;
+                if (Interop.Sys.IoSetup(EventBufferCount, &aioContext) == 0)
+                {
+                    _aioEvents = (Interop.Sys.IoEvent*)Marshal.AllocHGlobal(sizeof(Interop.Sys.IoEvent) * EventBufferCount);
+                    _aioBlocks = (Interop.Sys.IoControlBlock*)Marshal.AllocHGlobal(sizeof(Interop.Sys.IoControlBlock) * EventBufferCount);
+                    _aioBlocksPointers = (Interop.Sys.IoControlBlock**)Marshal.AllocHGlobal(sizeof(Interop.Sys.IoControlBlock*) * EventBufferCount);
+
+                    // Marshal.AllocHGlobal allocated memory might contain some garbage
+                    new Span<Interop.Sys.IoEvent>(_aioEvents, EventBufferCount).Clear();
+                    new Span<Interop.Sys.IoControlBlock>(_aioBlocks, EventBufferCount).Clear();
+
+                    for (int i = 0; i < EventBufferCount; i++)
+                    {
+                        _aioBlocksPointers[i] = &_aioBlocks[i];
+                    }
+
+                    _aioContext = aioContext;
+                }
+                else
+                {
+                    Debug.Assert(Interop.Sys.IsAioSupported() == false, "When AIO is supported IoSetup should always succeed");
```

Comment:
important: IoSetup might fail - on MacOS the API is simply missing, on Linux the user might not have enough of resources for the method to succeed